[Feature] Item tiers and new skills (Onslaught, Ruse and Momentum) from 12.80

omeranha opened the pull request:
The forge system part will be completed in another pull request, and focused after I finish these!
Continuation of this pull request will be in the forge, follow the link: #543

## New Item Tiers and classifications
![image](https://user-images.githubusercontent.com/53233626/166083422-8177bf62-f8df-4997-831a-b18b7b389c9b.png)
![image](https://user-images.githubusercontent.com/53233626/166083431-0683f70e-4929-435e-9475-e62fde55ffb9.png)
Added functions
item:getTier()
item:addTier(tier)

need a tiered item?
use the commited script, you can copy to OTservBR-Global datapack if you want

## New skills
All informations avaliable in Tibia Fandom
Onslaught(Fatal): https://tibia.fandom.com/wiki/Onslaught
Ruse(block hits): https://tibia.fandom.com/wiki/Ruse
Momentum(cooldown reduction): https://tibia.fandom.com/wiki/Momentum

## Missing
- [x] Missing some item tier bytes.
- [x] Market saving tier attribute.
and more, much more things.

## Type of change
  - [X] New feature (non-breaking change which adds functionality)

omeranha committed:
first commit

Zbizu commented on src/items/item.h:
```diff
@@ -1096,6 +1097,21 @@ class Item : virtual public Thing
+		}
+		void addTier(uint8_t tier) {
+			if (items[id].upgradeClassification) {
+				setIntAttr(ITEM_ATTRIBUTE_TIER, tier);
```
you're setting a tier, not adding it to existing value so a better name for this function would be `setTier`

omeranha replied:
Good observation, thank you!

Zbizu commented on src/items/items_definitions.hpp:
```diff
@@ -242,7 +242,8 @@ enum AttrTypes_t {
 	ATTR_QUICKLOOTCONTAINER = 38,
-	ATTR_IMBUEMENT_TYPE = 39
+	ATTR_IMBUEMENT_TYPE = 39,
+	ATTR_TIER = 40
```
if you set that to 41, you can use my RME to set tiers through properties window

omeranha replied:
hi zbizu, you mean set item tier in map using rme? I don't see the point

Zbizu commented on src/lua/functions/items/item_functions.cpp:
```diff
@@ -799,3 +799,31 @@ int ItemFunctions::luaItemGetImbuementSlot(lua_State* L) {
+}
+
+int ItemFunctions::luaItemAddTier(lua_State* L) {
+	// item:addTier(tier)
```
`setTier` recommended

Zbizu commented on src/lua/functions/items/item_functions.hpp:
```diff
@@ -85,6 +85,9 @@ class ItemFunctions final : LuaScriptInterface {
 			registerMethod(L, "Item", "getImbuement", ItemFunctions::luaItemGetImbuement);
 
+			registerMethod(L, "Item", "getTier", ItemFunctions::luaItemGetTier);
+			registerMethod(L, "Item", "addTier", ItemFunctions::luaItemAddTier);
```
`setTier` recommended

Zbizu commented on src/lua/functions/items/item_functions.hpp:
```diff
@@ -144,6 +147,9 @@ class ItemFunctions final : LuaScriptInterface {
 		static int luaItemGetImbuement(lua_State* L);
+		
+		static int luaItemGetTier(lua_State* L);
+		static int luaItemAddTier(lua_State* L);
```
`setTier` recommended

Zbizu commented on src/server/network/protocol/protocolgame.cpp:
```diff
@@ -4202,7 +4202,7 @@ void ProtocolGame::sendMarketAcceptOffer(const MarketOfferEx &offer)
 	msg.add<uint16_t>(offer.itemId);
 	if (Item::items[offer.itemId].upgradeClassification > 0) {
-		msg.addByte(0);
+		msg.addByte(Item::Item(offer.itemId).getTier());
```
you're making a new item here and calling a getter, this will always return 0
I suggest storing tiers as offer.tier

Zbizu commented on src/utils/utils_definitions.hpp:
```diff
@@ -764,4 +764,56 @@ const std::unordered_map<Blessings_t, std::string> BlessingNames = {
+	TIER10 = 10
+};
+
+const std::unordered_map<Tiers_t, double_t> dodgeChance = {
```
these can actually be calculated by a single formula, sample:
```lua
do
	local function calculateTierBonus(tier, x, y)
		return x * tier + y * (tier-1)^2
	end

	function ItemType:getTierBonus(tier)
		if tier == 0 then
			return -1, 0
		end
		
		if self:isWeapon() then
			-- onslaught (fatal)
			return SPECIALSKILL_ONSLAUGHT, calculateTierBonus(tier, 0.5, 0.05)
		elseif self:isArmor() then
			-- ruse (dodge)
			return SPECIALSKILL_RUSE, calculateTierBonus(tier, 0.5, 0.03)
		elseif self:isHelmet() then
			-- momentum (cooldown reset)
			return SPECIALSKILL_MOMENTUM, calculateTierBonus(tier, 2, 0.05)
		end
		
		return -1, 0
	end
end
```

rusaskii commented:
Global tibia examples:

You see a lion rod (magic level +2, critical hit chance 10%, critical extra damage +35%).
Imbuements: (Powerful Void 1:24h, Powerful Epiphany 1:24h).
Classification: 4 Tier: 1 (0.50% Onslaught).

Classification: 2 Tier: 1 (2.00% Momentum).

Classification: 3 Tier: 1 (0.50% Ruse).

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
code rework

omeranha committed:
should fix linux builds

omeranha committed:
sonar

omeranha committed:
math and sonar

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
fixes + forge system first commit (not completed yet)

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
progress saving

dudantas commented on src/items/item.cpp:
```diff
@@ -2286,6 +2301,20 @@ std::string Item::getDescription(const ItemType& it, int32_t lookDistance,
 
 	s << parseImbuementDescription(item);
 
+	if (uint16_t classification = item->getClassification(); classification > 1) {
```
You can extract this into a separated function, see example of parseImbuementDescription, and can pass item to reference for avoid nullpointer

Like this, function: 
```cpp
std::string Item::parseClassification(const Item& item) {
	std::ostringstream string;
	if (uint16_t classification = item.getClassification(); classification > 1) {
		string << std::endl << "Classification: " << classification << " Tier: " << item->getTier();
		if (item.getTier() != 0) {
			string  << " (";
			if (it.weaponType != WEAPON_NONE) {
				string << item.getFatalChance() << "% Onslaught).";
			} else if (g_game().getObjectCategory(item) == OBJECTCATEGORY_HELMETS) {
				string << item.getMomentumChance() << "% Momentum).";
			} else if (g_game().getObjectCategory(item) == OBJECTCATEGORY_ARMORS) {
				string << item.getDodgeChance() << "% Ruse).";
			}
		}
	}
	return string.str();
}
```

Call:
```cpp
parseClassification(*item);
```

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
sync + start fusion function

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
sync changes

omeranha committed:
commit set item tier talkaction

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
someday will be completed in another pull request

omeranha committed:
finished I think

omeranha committed:
ok

tutbarao commented:
BUG: Crash Server when looking at an NPC item

omeranha committed:
little fixes

tutbarao commented:
forged items can't be used as a hotkey

Zbizu commented:
> forged items can't be used as a hotkey

The packet got longer and includes u8 tier now.
Also you have to assign upgraded item again because the hotkey on action bar is set to equip "tier 0 armor"

tutbarao commented:
> > forged items can't be used as a hotkey
> 
> The packet got longer and includes u8 tier now. Also you have to assign upgraded item again because the hotkey on action bar is set to equip "tier 0 armor"

where should i change this to fix it?

Zbizu commented:
@tutbarao here is how I did it for hotkey equip:
https://github.com/Zbizu/forgottenserver/commit/2a18fc8ba813c2f38f87367fbc0e8350783a4f44#diff-eb01d79b279ebf659ed73216c0b3c490685221c9358e058ce55a1c8ab4e99c35R1145

I don't remember if there was anything I changed for hotkey use though.

tutbarao commented:
Thanks, is there any scripts that add dust loot drop chance on all monsters?

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
hotkey equip working but debug when having 1 tiered item and 1 normal item

omeranha committed:
fix debug, all working

wojciech-dybikowski commented:
when I click and select "sliver" or "exalted core"  client crashes with "2022-07-16, 23:15:02: Error while processing network packet not enough bytes (28777) available at position 25". Did I miss something or maybe you have the same issue?
![image](https://user-images.githubusercontent.com/85165190/179372078-901818ce-2847-45c1-8c88-1d8f7d611196.png)

omeranha committed:
sync

tutbarao commented:
Now the tier no longer appears
when looking at the item it shows tier = [] and the hotkey is only functional for tier 1 item.
above tier 1 it doesn't work

omeranha commented:
> Now the tier no longer appears when looking at the item it shows tier = [] and the hotkey is only functional for tier 1 item. above tier 1 it doesn't work

stop using the comment section as custom support, the pull request is in draft, that's = to not completed yet
please wait for everything to be done then you can test, I'll mark your comments as outdated

omeranha commented:
> when I click and select "sliver" or "exalted core" client crashes with "2022-07-16, 23:15:02: Error while processing network packet not enough bytes (28777) available at position 25". Did I miss something or maybe you have the same issue? ![image](https://user-images.githubusercontent.com/85165190/179372078-901818ce-2847-45c1-8c88-1d8f7d611196.png)

working as should there!

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
sonar, removed unecessary function and hotkey equip 100% working

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
sync

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
fixes

tutbarao commented:
When unequipping with hotkey the item with an infinite bag "loot pouch" with +1 page. Item disappears when unequipped. This bug happens with any item, even non-forged items.

carlospess0a commented:
Would be great if we have the getClassification() method in lua, what do you think?
**item_functions.cpp**

> int ItemFunctions::luaItemGetClassification(lua_State* L) {
> 	const Item* item = getUserdata<Item>(L, 1);
> 	if (!item) {
> 		reportErrorFunc(getErrorDesc(LUA_ERROR_ITEM_NOT_FOUND));
> 		pushBoolean(L, false);
> 		return 1;
> 	} 
> 
> 	lua_pushnumber(L, item->getClassification());
> 	return 1;
> }



**item_functions.hpp**

> 
> registerMethod(L, "Item", "getClassification", ItemFunctions::luaItemGetClassification);`
> 
> and
> 
> `static int luaItemGetClassification(lua_State* L);

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
item:getClassification function to lua

omeranha committed:
missing statistics

luiscaldera commented:
Hello, I just downloaded the current src with version 12.86. I added all the changes posted here. I tried several times to be sure that I did not make a mistake when adding them and it causes me an error when I try to log in 1 character, the server closes and only shows 1 message in the console (segmentation fault) the problem in windown and now in ubuntu 20.04

![image](https://user-images.githubusercontent.com/35881739/185739305-f47fe1d3-5bc5-49de-a5fa-80a11fa35558.png)

carlospess0a commented:
> outra coisa é que alguns items como "spellbook of dark mysteries" por algum motivo não para ser add ao market

Did you check your client with Asset Editor?

Luccoo commented:
Não sei se aqui é o lugar certo, dois erros nesse pr é, só da para criar offerts no market de itens que podem receber tier, e o outro só da para equipar quiver pela hotkey se não tiver nada na mão, por favor o pr é muito bom para ter esses dois bugs

edinh0xd reviewed (changes requested):
protocolgame.cpp
linha 4395
![image](https://user-images.githubusercontent.com/53832602/189379834-6d2ebc28-dd93-4245-b0c5-e45d6ae0c869.png)


	Item* slotItem = player->getInventoryItem(CONST_SLOT_BACKPACK);
	if (slotItem) {
		for (Item* item : slotItem ->getContainer()->getItems(true)) {
			if (item->getID() == ITEM_FORGE_SLIVER) {
				sliverCount += Item::countByType(item, -1);
			}
			else if (item->getID() == ITEM_FORGE_CORE) {
				coreCount += Item::countByType(item, -1);
			}
		}
	}

Luccoo commented:
quando um item com tier é vendido no market, ele perde o tier

omeranha committed:
Merge branch 'main' into item-tiers

omeranha committed:
fixing crash with characters without mainBp and clean

omeranha committed:
sonar

dudantas committed:
Merge branch 'main' into item-tiers

omeranha committed:
fix market attr

Luccoo commented:
os items não perdem mais tier quando são vendidos no market
mais ainda não é possível criar oferta de items que tem NÃO POSSUEM classificação tier

tutbarao commented:
same bug with me

dudantas committed:
Merge branch 'main' into item-tiers

dudantas committed:
Fix conflicts

dudantas committed:
Fix error

dudantas commented:
> os items não perdem mais tier quando são vendidos no market mais ainda não é possível criar oferta de items que tem NÃO POSSUEM classificação tier

que não possuem classificação tier ou que não possuem o upgradeclassfication? 
eu consegui por um item que tem tier 0 sem problema algum

tutbarao commented:
> > os items não perdem mais tier quando são vendidos no market mais ainda não é possível criar oferta de items que tem NÃO POSSUEM classificação tier
> 
> que não possuem classificação tier ou que não possuem o upgradeclassfication? eu consegui por um item que tem tier 0 sem problema algum

Qualquer item que nao tenha a classificação tier nao pode ser adicionado ao market. Exemplo: steel boots nao pode ser adicionado ao market pois nao possue tier.

dudantas committed:
Some tier fixes

Fixed only items with classification 0 appearing on market
Fixed issue with non-tiered items not selling on the market
Fixexd iomarket tier size
Fixed logs from offers on create and accept market offers

dudantas committed:
Fix tier order

dudantas commented:
> Qualquer item que nao tenha a classificação tier nao pode ser adicionado ao market. Exemplo: steel boots nao pode ser adicionado ao market pois nao possue tier.

Acabei de arrumar, fica a vontade para testar. Também botei para aparecer todas as classificações no market, na hora de pôr para vender.

dudantas committed:
Fix migrations

dudantas commented on src/io/iomarket.cpp:
```diff
@@ -147,6 +151,10 @@ void IOMarket::processExpiredOffers(DBResult_ptr result, bool)
 				while (tmpAmount > 0) {
 					uint16_t stackCount = std::min<uint16_t>(100, tmpAmount);
 					if (Item* item = Item::CreateItem(itemType.id, stackCount); g_game().internalAddItem(player->getInbox(), item, INDEX_WHEREEVER, FLAG_NOLIMIT) != RETURNVALUE_NOERROR) {
+						if (tier != 0) {
```
This is wrong, it will only run if there is a problem creating the item

dudantas commented on src/io/iomarket.cpp:
```diff
@@ -164,6 +172,10 @@ void IOMarket::processExpiredOffers(DBResult_ptr result, bool)
 				for (uint16_t i = 0; i < amount; ++i) {
 					Item* item = Item::CreateItem(itemType.id, subType);
 					if (g_game().internalAddItem(player->getInbox(), item, INDEX_WHEREEVER, FLAG_NOLIMIT) != RETURNVALUE_NOERROR) {
+						if (tier != 0) {
```
Same

dudantas committed:
Some fix

Fixed setIntAttr order and others things

dudantas committed:
Fix uint size

tutbarao commented:
Eu tenho esse erro agora na source ao criar uma oferta no market de uma steel boots ou qualquer outro item sem tier.

```sql
`[error] Query: SELECT `id`, `amount`, `price`, `tier`, `created`, `anonymous`, (SELECT `name` FROM `players` WHERE `id` = `player_id`) AS `player_name` FROM `market_offers` WHERE `sale` = 0 AND `itemtype` = 3554 AND `tier` =
[2022-06-10 17:48:33.312] [error] Message: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1
[2022-06-10 17:48:33.313] [error] Query: SELECT `id`, `amount`, `price`, `tier`, `created`, `anonymous`, (SELECT `name` FROM `players` WHERE `id` = `player_id`) AS `player_name` FROM `market_offers` WHERE `sale` = 0 AND `itemtype` = 3554 AND `tier` =
[2022-06-10 17:48:33.314] [error] Message: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1
[2022-06-10 17:48:33.315] [error] Query: SELECT `id`, `amount`, `price`, `tier`, `created`, `anonymous`, (SELECT `name` FROM `players` WHERE `id` = `player_id`) AS `player_name` FROM `market_offers` WHERE `sale` = 1 AND `itemtype` = 3554 AND `tier` =
[2022-06-10 17:48:33.315] [error] Message: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1
[2022-06-10 17:48:33.316] [error] Query: SELECT `id`, `amount`, `price`, `tier`, `created`, `anonymous`, (SELECT `name` FROM `players` WHERE `id` = `player_id`) AS `player_name` FROM `market_offers` WHERE `sale` = 1 AND `itemtype` = 3554 AND `tier` =
[2022-06-10 17:48:33.316] [error] Message: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1`
```

dudantas commented:
@tutbarao você precisa rodar o migrations
https://github.com/opentibiabr/otservbr-global/pull/799

#Edit
Realmente ta dando bug, já já vejo oq é

dudantas committed:
Fix tier convert to string

dudantas commented:
> Eu tenho esse erro agora na source ao criar uma oferta no market de uma steel boots ou qualquer outro item sem tier.
> 
> ```sql
> `[error] Query: SELECT `id`, `amount`, `price`, `tier`, `created`, `anonymous`, (SELECT `name` FROM `players` WHERE `id` = `player_id`) AS `player_name` FROM `market_offers` WHERE `sale` = 0 AND `itemtype` = 3554 AND `tier` =
> [2022-06-10 17:48:33.312] [error] Message: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1
> [2022-06-10 17:48:33.313] [error] Query: SELECT `id`, `amount`, `price`, `tier`, `created`, `anonymous`, (SELECT `name` FROM `players` WHERE `id` = `player_id`) AS `player_name` FROM `market_offers` WHERE `sale` = 0 AND `itemtype` = 3554 AND `tier` =
> [2022-06-10 17:48:33.314] [error] Message: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1
> [2022-06-10 17:48:33.315] [error] Query: SELECT `id`, `amount`, `price`, `tier`, `created`, `anonymous`, (SELECT `name` FROM `players` WHERE `id` = `player_id`) AS `player_name` FROM `market_offers` WHERE `sale` = 1 AND `itemtype` = 3554 AND `tier` =
> [2022-06-10 17:48:33.315] [error] Message: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1
> [2022-06-10 17:48:33.316] [error] Query: SELECT `id`, `amount`, `price`, `tier`, `created`, `anonymous`, (SELECT `name` FROM `players` WHERE `id` = `player_id`) AS `player_name` FROM `market_offers` WHERE `sale` = 1 AND `itemtype` = 3554 AND `tier` =
> [2022-06-10 17:48:33.316] [error] Message: You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '' at line 1`
> ```

Fixado, vê o último commit, por gentileza.

![image](https://user-images.githubusercontent.com/8551443/194437270-1c9b67c4-15f2-4a1d-a6e4-375ebe4706a4.png)

dudantas committed:
Merge branch 'main' into item-tiers

dudantas commented:
Mais algum bug? Precisamos que testem bem, amanhã começarei a desenvolver a statistics e aí estará pronto para mergear. Só dependerá dos testes.

dudantas committed:
Some indent

dudantas committed:
Some indent

dudantas committed:
Fix

dudantas commented on src/items/item.cpp:
```diff
@@ -1502,6 +1521,24 @@ std::string Item::parseImbuementDescription(const Item* item)
 
+std::string Item::parseClassificationDescription(const Item* item) {
+	std::ostringstream string;
+	if (item && item->getClassification() >= 1) {
```
> `if (item && item->getClassification() >= 0) {`
> 
> To show any item without tier classification.

I believe that only item rated 1 or more should be shown.

dudantas committed:
Fix

dudantas committed:
Some item tier fixes

dudantas committed:
Small fix to tier and upgrade detail

dudantas committed:
Add market purchase and sale statistics for item tiers

dudantas committed:
Fix statistics values and some tier changes

Created "local" function on iomarket for get tier value from database
Removed getPurchaseStatistics and getSaleStatistics functions, changed to return only the map, map is read on protocolgame
Fixed get and set tier functions on item.h for check if tier is valid
Created MAX_ITEM_FORGE_TIER variable to define the max item tier

Changed ProtocolGame::sendMarketDetail to read new statistics functions

dudantas committed:
Some tier fixes

Fixed formula from fatal dodge
Fixed depot search item
Removed log from IOMarket::updateStatistics

dudantas committed:
Some tier fixes and sonar adjustments

Fixed fatal dodge formula
Fixed request depot search item tier
Some sonar check fixes

guispiller commented:
Everything working perfect, i would just suggest that tiered items should not be able to be sold on npcs

beats-dh committed:
fix buid linux, sonar

dudantas committed:
Fix for not show item with imbuement and tier on npc shop

Fixed some npc shop functions and fixed market price size from uint32_t to uint64_t

dudantas committed:
Merge branch 'item-tiers' of https://github.com/opentibiabr/canary into item-tiers

dudantas committed:
Fix sell offer remove item

dudantas committed:
Some sonar check fixes and others things

Fixed item tier update on item creation with "/i" command
Indented some codes and fixes others things

dudantas committed:
Fix create item tier

dudantas committed:
Fix to remove only the selected items count

dudantas committed:
Add auto instead of types

dudantas committed:
Fix possibility of clone money

Npc could give money even if no items are removed

dudantas committed:
Simplify functions body and rename function for better explanation

Renamed function from "getAllItemTypeCount" to "getAllSaleItemIdAndCount"
Simplified the functions of get inventory items

dudantas committed:
Fix sell items

Fixed sell items to npc and add money logic

guispiller commented:
Found a new issue on market related to tiers. How to reproduce, with an example of Plate armor (you can use any item).

1 - create a buy offer of a plate armor tier 0 with Player A.
2 - Get 1 Plate armor Tier 1 on first slot of a depot box, and one plate armor Tier 0 at the second slot of the same depot box.
3 - Sell the plate armor tier 0 to the buy offer at the market. 

Result: As the first slot is the Tier 1 armor, this one will be sold. So, player A will receive a Tier 0 armor, and player B will sell a Tier 1 armor.
This can occur also if there are tiered items of the same at mailbox, and lower tiered on depot box. It is not checking the tier, it's just selling the first one

dudantas committed:
Fix item tier ignore from requestLockerItems

dudantas commented:
> Found a new issue on market related to tiers. How to reproduce, with an example of Plate armor (you can use any item).
> 
> 1 - create a buy offer of a plate armor tier 0 with Player A. 2 - Get 1 Plate armor Tier 1 on first slot of a depot box, and one plate armor Tier 0 at the second slot of the same depot box. 3 - Sell the plate armor tier 0 to the buy offer at the market.
> 
> Result: As the first slot is the Tier 1 armor, this one will be sold. So, player A will receive a Tier 0 armor, and player B will sell a Tier 1 armor. This can occur also if there are tiered items of the same at mailbox, and lower tiered on depot box. It is not checking the tier, it's just selling the first one

Thanks for reporting. Please, test this commit: https://github.com/opentibiabr/canary/pull/366/commits/d0d3482da484f1beb9d3966a65210fc4faebf26a

I tested it, and it seems to have fixed the problem.

dudantas committed:
Some sonar fixes

tutbarao commented:
I found a bug: Cannot cancel market offers. When you click the cancel button, nothing happens.

tutbarao commented:
![image](https://user-images.githubusercontent.com/28155884/197549281-c01f745a-eacc-4228-955d-4dea307bb570.png)
I got this error when starting the server. Perhaps this is the cause of the above bug.

dudantas committed:
Fix tier error on distro

dudantas commented:
> ![image](https://user-images.githubusercontent.com/28155884/197549281-c01f745a-eacc-4228-955d-4dea307bb570.png) I got this error when starting the server. Perhaps this is the cause of the above bug.

Fixed with last commit (https://github.com/opentibiabr/canary/pull/366/commits/ec64bb541aa04c6919d402dc1899d8614b1afcd0), thanks for reporting

dudantas committed:
Remove comment

dudantas reviewed (approved):
Pull finished, it was tested in several scenarios, by several people, what was reported has already been fixed and now we can merge.

dudantas commented:
> Tem que ver se não vai crashar / debugar combinando: Critical + Fatal **(tier)** Dodge **(tier)** + Critical
> 
> Durante battle com monstros que refletem dano ou absorvam dano. ex; Burster Spectre ou Bosses refletindo _ao mesmo tempo_ em que o **tier gera** um possivel **dodge**. Com imbuements ou Runas de Aumento de mana/life leech e/ou Critico (Low Blow)
> 
> Fora outras combinações que não me vem a cabeça agora 😵‍💫

Nós já estamos trabalhando no pull request há um bom tempo e resolvemos tudo que foi reportado e o que encontramos de bug. Se você quiser testar este cenário e informar se deu algum problema, é de bom grado. Particularmente eu não tenho interesse em realizar este teste. Caso tenha interesse, é só reportar no pull request da forja, este já vai para o merge...

dudantas commented:
Continuation of this pull request will be in the forge, follow the link: #543
